File: platforms/android/capture-runtime/src/main/java/com/tugboat/capture/internal/MaskMapper.kt
package com.tugboat.capture.internal

internal object MaskMapper {
    fun toPixelRects(
        masks: List<com.tugboat.capture.NormalizedMask>,
        width: Int,
        height: Int,
    ): IntArray {
        if (width <= 0 || height <= 0) return IntArray(0)
        val packed = ArrayList<Int>(masks.size * 4)
        for (mask in masks) {
            if (mask.width <= 0.0 || mask.height <= 0.0) continue
            val left = kotlin.math.floor(mask.x * width).toInt().coerceIn(0, width)
            val top = kotlin.math.floor(mask.y * height).toInt().coerceIn(0, height)
            val right =
                kotlin.math.ceil((mask.x + mask.width) * width).toInt().coerceIn(0, width)
            val bottom =
                kotlin.math.ceil((mask.y + mask.height) * height).toInt().coerceIn(0, height)
            if (right <= left || bottom <= top) continue
            packed.add(left)
            packed.add(top)
            packed.add(right)
            packed.add(bottom)
        }
        return packed.toIntArray()
    }
}

File: platforms/apple/Sources/TugboatCaptureRuntime/Internal/MaskMapper.swift
import Foundation

enum MaskMapper {
  static func toPixelRects(masks: [NormalizedMask], width: Int, height: Int) -> [Int32] {
    if width <= 0 || height <= 0 {
      return []
    }
    var packed: [Int32] = []
    packed.reserveCapacity(masks.count * 4)
    for mask in masks {
      if mask.width <= 0.0 || mask.height <= 0.0 {
        continue
      }
      let left = clamp(Int(floor(mask.x * Double(width))), 0, width)
      let top = clamp(Int(floor(mask.y * Double(height))), 0, height)
      let right = clamp(Int(ceil((mask.x + mask.width) * Double(width))), 0, width)
      let bottom = clamp(Int(ceil((mask.y + mask.height) * Double(height))), 0, height)
      if right <= left || bottom <= top {
        continue
      }
      packed.append(Int32(left))
      packed.append(Int32(top))
      packed.append(Int32(right))
      packed.append(Int32(bottom))
    }
    return packed
  }

  private static func clamp(_ value: Int, _ lower: Int, _ upper: Int) -> Int {
    min(max(value, lower), upper)
  }
}
